src/Backend/SgateShopgatePlugin/Controllers/Frontend/shopgate.php

```diff
@@ -754,6 +785,109 @@ public function updateCartItemAction()
         exit();
     }
 
+    /**
+     * Controller action for login error message in frontend
+     */
+    public function errorAction()
+    {
+        Shopware()->Events()->notify(
+            'Shopgate_Frontend_Custom_Event',
+            array(
+                'request' => $this->Request(),
+                'subject' => $this
+            )
+        );
+    }
+
+    /**
+     * Login app user from JWT token
+     *
+     * @param $token
+     * @return bool
+     */
+    protected function loginAppUser($token)
+    {
+        $basket = $this->basket->sGetBasket();
+        $voucher = $this->getVoucher();
+
+        if (isset($token)) {
+            $key = trim($this->getConfig()->getApikey());
+            JWT::$leeway = 60;
+            $decoded = JWT::decode($token, $key, array('HS256'));
+            $decoded = json_decode(json_encode($decoded), true);
+            $customerId = $decoded['customer_id'];
+            $promotionVouchers = json_decode($decoded['promotion_vouchers'], true);
+
+            if (isset($promotionVouchers)) {
+                $this->session->offsetSet('promotionVouchers', $promotionVouchers);
+            }
+
+            $sql = 'SELECT DISTINCT `password` FROM `s_user` WHERE customernumber=?';
+            $password = Shopware()->Db()->fetchCol($sql, array($customerId));
+
+            $sql = 'SELECT DISTINCT `email` FROM `s_user` WHERE customernumber=?';
+            $email = Shopware()->Db()->fetchCol($sql, array($customerId));
+
+            if (count($password) > 1 || count($email) > 1) {
+                return false;
+            }
+
+            $this->Request()->setPost('email', $email[0]);
```

Comment:
duplicate block of code, see function getUserAction(), line 1081-1090. Outsource to a function or helper.